library/analytics/src/main/java/org/cru/godtools/analytics/firebase/FirebaseAnalyticsService.kt

```diff
@@ -29,13 +31,20 @@ private const val VALUE_APP_TYPE_INSTANT = "instant"
 private const val VALUE_APP_TYPE_INSTALLED = "installed"
 
 @Singleton
-class FirebaseAnalyticsService @MainThread @Inject internal constructor(
+class FirebaseAnalyticsService @VisibleForTesting internal constructor(
     app: Application,
     eventBus: EventBus,
-    oktaUserProfileProvider: OktaUserProfileProvider
+    oktaUserProfileProvider: OktaUserProfileProvider,
+    private val firebase: FirebaseAnalytics,
+    coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.Default)
```

Comment:
Did you have to add the coroutineScope for testing purposes?

Reply:
yeah, I had to expose it so that I could provide a testing coroutineScope

Reply:
there is a coroutines-test library that provides a `TestCoroutineScope`, this test scope provides the following:
- coroutines execute immediately and the original call doesn't return until the coroutine is complete
- full control over the execution clock
- methods to ensure that there are no pending coroutines once your test is complete

Reply:
the first capability is the critical one here, without it we wouldn't be able to immediately test that code in coroutines ran, we would have to introduce an artificial delay into our test and hope that the code runs before the delay expires.

Reply:
here's an example test utilizing control over the execution clock: https://github.com/CruGlobal/godtools-android/blob/develop/ui/tract-renderer/src/test/java/org/cru/godtools/tract/ui/controller/HeroControllerTest.kt#L60-L89